src/tests/testSBuf.cc

```diff
@@ -850,6 +850,33 @@ testSBuf::testReserve()
         buffer.reserve(requirements);
         CPPUNIT_ASSERT(buffer.spaceSize() >= requirements.minSpace);
     }
+
+    { // MemBlob move-to-start operation after chop
+        SBuf b(fox);
+        b.chop(5);
+        b.reserveSpace(5);
+        CPPUNIT_ASSERT(b.spaceSize() >= 5);
+        SBuf ref(fox + 5);
+        CPPUNIT_ASSERT_EQUAL(ref,b);
+    }
+
+    { // MemBlob crop operations after chop
+        SBuf b(fox);
+        b.chop(5, 10);
+        const auto space = static_cast<SBuf::size_type>(5) + b.spaceSize();
+        b.reserveSpace(space);
+        CPPUNIT_ASSERT(b.spaceSize() >= space);
+    }
```

Comment:
Fixed with commit  b7a0810ab7

Reply:
The new test cases look much better, thank you. I do not request further test case changes right now, but before I can polish them further, I have one question: You are using raw pointers to check that there was no reallocation. Technically, reallocation may result in the same raw buffer space being allocated, especially if there are two reallocations happening under the hood. Would it be possible to rely on SBuf _statistics_ (i.e. SBufStats) rather than (or in addition to?) raw pointer equivalence to detect desirable operations?

Reply:
> Technically, reallocation may result in the same raw buffer space being allocated ...

This is not possible for SBuf. When SBuf "re-allocates" in practice allocates a new blob while the old blob is still  used. Then we are copying the data from old blob to the new and then SBuf removes the reference to the old blob.

Also a similar check is also used in other place (testSBuf::testGrow) to check reallocation.  

> Would it be possible to rely on SBuf statistics (i.e. SBufStats) rather than (or in addition to?) raw pointer equivalence to detect desirable operations?

We can check statistics, but also the statistics may lie(because of a bug). The raw addresses tests are accurate. The statistics may have its own tests.

Reply:
> This is not possible for SBuf. When SBuf "re-allocates" in practice allocates a new blob while the old blob is still  used.

It is definitely possible with multiple internal reallocations -- the first "old blob" may be gone by the time the second reallocation starts. The tests do not count reallocations so they cannot be sure that calling non-constant SBuf methods will not result in multiple reallocations, especially when calling multiple non-constant SBuf methods in a buggy SBuf code (that the tests are trying to detect).

> the statistics may lie

True, but raw pointers are not reliable either (as discussed above). We could use both metrics at the same time, of course. Violating at least one check would be enough to declare a test failure.

Reply:
commit ef1fc9f8